src/Microsoft.OData.Core/JsonLight/ODataJsonLightEntityReferenceLinkDeserializer.cs

```diff
@@ -465,5 +465,328 @@ private ODataEntityReferenceLink ReadSingleEntityReferenceLink(PropertyAndAnnota
+        ///                 JsonNodeType.EndObject              The end of the entity reference links object
+        /// Post-Condition: JsonNodeType.EndObject              When the end of the entity reference links object is reached
+        ///                 Any                                 The first node of the value of the 'url' property (if found)
+        /// </remarks>
+        private async Task ReadEntityReferenceLinksAnnotationsAsync(
+            ODataEntityReferenceLinks links,
+            PropertyAndAnnotationCollector propertyAndAnnotationCollector,
+            bool forLinksStart)
+        {
+            Debug.Assert(links != null, "links != null");
+            Debug.Assert(propertyAndAnnotationCollector != null, "propertyAndAnnotationCollector != null");
+            this.AssertJsonCondition(JsonNodeType.Property, JsonNodeType.EndObject);
+            this.JsonReader.AssertNotBuffering();
+
+            while (this.JsonReader.NodeType == JsonNodeType.Property)
+            {
+                // OData property annotations are not supported on entity reference links.
+                Func<string, Task<object>> propertyAnnotationValueReaderDelegate =
+                    annotationName =>
+                    {
+                        throw new ODataException(ODataErrorStrings.ODataJsonLightEntityReferenceLinkDeserializer_PropertyAnnotationForEntityReferenceLinks);
+                    };
+
+                bool foundValueProperty = false;
+                this.ReadPropertyCustomAnnotationValueAsync = this.ReadCustomInstanceAnnotationValueAsync;
+                await this.ProcessPropertyAsync(
+                    propertyAndAnnotationCollector,
+                    propertyAnnotationValueReaderDelegate,
+                    async (propertyParseResult, propertyName) =>
+                    {
+                        if (this.JsonReader.NodeType == JsonNodeType.Property)
+                        {
+                            // Read over property name
+                            await this.JsonReader.ReadAsync()
+                                .ConfigureAwait(false);
+                        }
+
+                        switch (propertyParseResult)
+                        {
+                            case PropertyParsingResult.ODataInstanceAnnotation:
+                                if (string.Equals(ODataAnnotationNames.ODataNextLink, propertyName, StringComparison.Ordinal))
+                                {
+                                    this.AssertJsonCondition(JsonNodeType.PrimitiveValue);
+                                    Uri nextPageUri = await this.ReadAndValidateAnnotationStringValueAsUriAsync(ODataAnnotationNames.ODataNextLink)
+                                        .ConfigureAwait(false);
+                                    Debug.Assert(links.NextPageLink == null, "We should have checked for duplicates already.");
+                                    links.NextPageLink = nextPageUri;

```

Comment:
The code in that method is trivial so I considered it prudent to have that logic within this case block instead of incurring the extra overhead of a state machine by introducing an equivalent async method. The logic is equivalent.